src/dmn/trn_transit_xdp_usr.c

```diff
@@ -645,6 +727,82 @@ int trn_user_metadata_init(struct user_metadata_t *md, char *itf,
 	return 0;
 }
 
+int trn_user_metadata_init_offload(struct user_metadata_t *md, char *itf,
+			   char *kern_path, int xdp_flags)
+{
+	int rc;
+	struct rlimit r = { RLIM_INFINITY, RLIM_INFINITY };
+	struct bpf_prog_load_attr prog_load_attr = { .prog_type =
+							     BPF_PROG_TYPE_XDP,
+						     .file = kern_path };
+	__u32 info_len = sizeof(md->info_offload);
+
+	if (setrlimit(RLIMIT_MEMLOCK, &r)) {
+		TRN_LOG_ERROR("setrlimit(RLIMIT_MEMLOCK)");
+		return 1;
+	}
+
+	md->ifindex = if_nametoindex(itf);
+	prog_load_attr.ifindex = md->ifindex;
+	if (!md->ifindex) {
+		TRN_LOG_ERROR("if_nametoindex");
+		return 1;
+	}
+
+	md->eth.ip = trn_get_interface_ipv4(md->ifindex);
+	md->eth.iface_index = md->ifindex;
+
+	// offload_xdp cannot reuse the pinned maps(network policy)
+	if (bpf_prog_load_xattr(&prog_load_attr, &md->obj_offload, &md->prog_offload_fd)) {
+		TRN_LOG_ERROR("Error loading bpf: %s", kern_path);
+		return 1;
+	}
+
+	// map_init
+	md->networks_offload_map = bpf_map__next(NULL, md->obj_offload);
+	md->vpc_offload_map = bpf_map__next(md->networks_offload_map, md->obj_offload);
+	md->endpoints_offload_map = bpf_map__next(md->vpc_offload_map, md->obj_offload);
+	md->interface_config_offload_map = bpf_map__next(md->endpoints_offload_map, md->obj_offload);
+	if (!md->endpoints_offload_map || !md->interface_config_offload_map ||
+		!md->networks_offload_map || !md->vpc_offload_map) {
+		TRN_LOG_ERROR("Failure finding offloaded maps objects.");
+		return 1;
+	}
+	md->networks_offload_map_fd = bpf_map__fd(md->networks_offload_map);
+	md->vpc_offload_map_fd = bpf_map__fd(md->vpc_offload_map);
+	md->endpoints_offload_map_fd = bpf_map__fd(md->endpoints_offload_map);
+	md->interface_config_offload_map_fd = bpf_map__fd(md->interface_config_offload_map);
+	// map_init done
+
+	if (!md->prog_offload_fd) {
+		TRN_LOG_ERROR("load_bpf_file: %s.", strerror(errno));
+		return 1;
+	}
+
+	if (bpf_set_link_xdp_fd(md->ifindex, md->prog_offload_fd, xdp_flags) < 0) {
+		TRN_LOG_ERROR("link set xdp_offload fd failed - %s.", strerror(errno));
+		return 1;
+	}
+
+	rc = bpf_obj_get_info_by_fd(md->prog_offload_fd, &md->info_offload, &info_len);
+	if (rc != 0) {
+		TRN_LOG_ERROR("can't get prog info - %s.", strerror(errno));
+		return rc;
+	}
+	md->prog_offload_id = md->info_offload.id;
+
+	// As the config of original Transit Program already has the itf_idx, set Offload Program as the same config
+	int k = 0;
+	rc = bpf_map_update_elem(md->interface_config_offload_map_fd, &k, &md->eth, 0);
+	if (rc != 0) {
+		TRN_LOG_ERROR("Failed to store interface data.");
+		return 1;
+	}
+
+	md->xdp_flags = xdp_flags;  // overwrite xdp_flags with XDP_OFFLOAD
```

Comment:
Does this interfere with operation of non-offloaded traffic?

Reply:
It does not interfere with operation of non-offloaded traffic.